### ISSUE TITLE
$inc on sub-document array sets and does not increment

### ISSUE THREAD

**Issue by dvetal-cigna** [mongomock/mongomock#428]
Problem:

If I have a document like ```{'name': 'bob', 'data': [0, 0]}``` and I wish to increment the 2nd
index on the 'data' array mongomock will not appropriately update the value.  In fact, it will $set on the data.1 location.

I've written a test below that can be injected into the existing test suite.
```
    def test__inc_subdocument_array(self):
        self.cmp.do.remove()
        self.cmp.do.insert({'name': 'bob', 'data': [0, 0]})
        self.cmp.do.update({'name': 'bob'}, {'$inc': {'data.1': 1}})
        self.cmp.compare.find()
        self.cmp.do.update({'name': 'bob'}, {'$inc': {'data.1': -3}})
        self.cmp.compare.find()
```
Currently the test will result in a document ```{'name': 'bob', 'data': [0, -3]}``` but I would expect the result to be ```{'name': 'bob', 'data': [0, -2]}```

p.s. I was attempting branch off and fix the issue myself for you guys but when I ran nosetests I was having issues. First, it was hard for me to identify where the test was being set to skip. Second, for some reason even if I put self.assertEqual(1, 0) or assert 1==0 in my test the test still continued to pass. Lastly, even using -nocapture and setting --pdb I could not get the system to allow me to pdb.set_trace() which I have never had a problem with in the past.   If you can tell me what I am doing wrong I'd be happy to push forward and get a MR in. However, I'm not even going to attempt unless I can be confident my test is actually running.

**Comment by pcorpet** [comment]
Hi @dvetal-cigna,

Thanks for your report and your attempts to solve the bug. It's indeed a bug in the code next to:

```py
          if isinstance(doc, list):
              try:
                  doc[int(field_name)] = field_value
              except IndexError:
                  pass
          else:
              updater(doc, field_name, field_value)
```

As you can see if the `doc` is a `list`, then the `updater` is ignored and the field value is set directly.

For running the tests: depending on what's installed locally different tests are run. I think I should update the doc, but I haven't done a lot of local testing myself and as such I'm not sure exactly the right pattern. I recommend that you install locally `pymongo`, `pyexecjs` and have a local MongoDB instance running on `localhost:27017` before running the tests. With that, most of the tests should be running (and hopefully passing). If you don't succeed, you can still create a PR and Travis will run the tests for you.

**Comment by dvetal-cigna** [comment]
@pcorpet I have a fix ready on a local branch. For some reason I can't push my branch. Getting 403. Do you have to add me as a collaborator?

**Comment by armudgal** [comment]
@dvetal-cigna, you have to first fork the repo, come your forked repo, make changes in it and then open a Pull request for the same

**Comment by pcorpet** [comment]
Thanks @dvetal-cigna for your fix in #429
